### ISSUE TITLE
SSH handshake failure not propagating

### ISSUE THREAD

**Issue by dluchinienoc** [mitchellh/packer#2004]
An SSH authentication error is not being propagated to the "Finished" status:

```
+ /usr/bin/packer build -only=amazon-ebs -color=false base.json

Running the 'amazon-ebs' builder task
amazon-ebs output will be in this color.

==> amazon-ebs: Inspecting the source AMI...
==> amazon-ebs: Creating temporary keypair: packer 5515b3e0-8e78-fc2f-7506-4b467f43e625
==> amazon-ebs: Launching a source AWS instance...
    amazon-ebs: Instance ID: i-3619b5e1
==> amazon-ebs: Waiting for instance (i-3619b5e1) to become ready...
==> amazon-ebs: Waiting for SSH to become available...
==> amazon-ebs: Error waiting for SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [publickey none], no supported methods remain
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: Deleting temporary keypair...
Build 'amazon-ebs' finished.

==> Builds finished. The artifacts of successful builds are:
Finished: SUCCESS
```

The fact that Packer isn't able to connect to the machine is a problem on our end, and is not a Packer problem in and of itself. The problem is that Packer isn't realizing that there was an error and is returning status code zero (in this case shown by Jenkins as `Finished: SUCCESS`).

Reproducible with Packer 0.7.5 on OS X 10.9 and CentOS 6.

**Comment by m1keil** [comment]
We are hitting this bug as well. In our case we have number of provision scripts running one after another and the drop of SSH just cause packer to start running the next provision script in line without notifying about error.

This is even reproducible locally with VirtualBox:
```
{
  "builders": [
    {
      "name": "virtualbox",
      "type": "virtualbox-ovf",
      "source_path": "/path/to/ovf",
      "vm_name": "packer_demo",
      "ssh_username": "vagrant",
      "ssh_key_path": "/path/to/insecure_private_key",
      "ssh_wait_timeout": "2m",
      "shutdown_command": "sudo -S shutdown -P now",
      "headless": true
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "script": "long_timeout.sh"
    },
    {
      "type": "shell",
      "script": "long_timeout.sh"
    }
 ]
}
```

`long_timeout.sh` is just a long sleep.

If you start `packer build` with this config, packer will wait for the first provision to stop. If you will initiate some network disconnection event (like shutting down WIFI or disconnecting the ethernet cable), this will cause SSH to drop and packer will just continue with the next provision in line:

```
1429127396,,ui,say,==> virtualbox: Connected to SSH!
1429127396,,ui,say,==> virtualbox: Uploading VirtualBox version info (4.3.26)
1429127397,,ui,say,==> virtualbox: Uploading VirtualBox guest additions ISO...
1429127401,,ui,say,==> virtualbox: Provisioning with shell script: long_timeout.sh
<------------ I disconnected wifi here ------------>
1429127432,,ui,say,==> virtualbox: Provisioning with shell script: long_timeout.sh
```

So in our case we don't even get the error about SSH drop in log.

**Comment by catsby** [comment]
I belive this issue to be resolved now. I used @m1keil 's example and I can't reproduce it:

```console
$ packer build quickstart.json
virtualbox output will be in this color.

==> virtualbox: Downloading or copying Guest additions
    virtualbox: Downloading or copying: file:///Applications/VirtualBox.app/Contents/MacOS/VBoxGuestAdditions.iso
==> virtualbox: Importing VM: /Users/clint/Desktop/boot2docker-extra-vm.ova
==> virtualbox: Creating forwarded port mapping for SSH (host port 2559)
==> virtualbox: Starting the virtual machine...
    virtualbox: WARNING: The VM will be started in headless mode, as configured.
    virtualbox: In headless mode, errors during the boot sequence or OS setup
    virtualbox: won't be easily visible. Use at your own discretion.
==> virtualbox: Waiting 10s for boot...
==> virtualbox: Typing the boot command...
==> virtualbox: Waiting for SSH to become available...
==> virtualbox: Error waiting for SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password keyboard-interactive publickey], no supported methods remain
==> virtualbox: Unregistering and deleting imported VM...
==> virtualbox: Deleting output directory...
Build 'virtualbox' errored: Build was halted.

==> Some builds didn't complete successfully and had errors:
--> virtualbox: Build was halted.

==> Builds finished but no artifacts were created.
```

@dluchinienoc can you confirm this is or is not still an issue? 

Thanks!

**Comment by m1keil** [comment]
@catsby, I'm still able to replicate my example under 0.7.5

**Comment by dluchinienoc** [comment]
I originally reported this in 0.7.5 and don't think there have been any commits against this issue; there also hasn't been a Packer release since.

Is it possible that the failure mode differs by builder? My example was with the `amazon-ebs` builder, but @m1keil's is for the `virtualbox-ovf` builder.

**Comment by catsby** [comment]
@m1keil 's example also fails in a different place; in that example, ssh connection is made:

```
1429127396,,ui,say,==> virtualbox: Connected to SSH!
1429127396,,ui,say,==> virtualbox: Uploading VirtualBox version info (4.3.26)
```

In mine (and I thought @dluchinienoc 's), the error was during ssh:

```
==> amazon-ebs: Error waiting for SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [publickey none], no supported methods remain
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: Deleting temporary keypair...
Build 'amazon-ebs' finished.

==> Builds finished. The artifacts of successful builds are:
Finished: SUCCESS
```

Do you recall, were there other artifacts involved there? 

> The problem is that Packer isn't realizing that there was an error and is returning status code zero 

This is the part that I can't reproduce :frowning: , but I'll keep trying ...

**Comment by m1keil** [comment]
@dluchinienoc any chance you can share with us the ami or at least if there were any sshd config changes from the default centos6 configs?

**Comment by catsby** [comment]
I'm fairly confident this issue is resolved in the current `master` branch. 
To test, I waited for Packer to enter the `Waiting for SSH to become available...` stage, and then turned off my WiFi. 

I used this config:

```javascript
{
  "builders": [
    {
      "type": "amazon-ebs",
      "region": "us-west-2",
      "source_ami": "ami-dfc39aef",
      "instance_type": "t2.micro",
      "ssh_username": "ec2-user",
      "ami_name": "packer-testing-{{timestamp}}",
      "tags": {
        "packer-test": true
      }
    }
  ]
}

```

which gave me these results:

```console
$ packer build quickstart.json
amazon-ebs output will be in this color.

==> amazon-ebs: Prevalidating AMI Name...
==> amazon-ebs: Inspecting the source AMI...
==> amazon-ebs: Creating temporary keypair: packer 5579e23c-e557-986a-0a19-c64e325a3982
==> amazon-ebs: Creating temporary security group for this instance...
==> amazon-ebs: Authorizing SSH access on the temporary security group...
==> amazon-ebs: Launching a source AWS instance...
    amazon-ebs: Instance ID: i-18397eef
==> amazon-ebs: Waiting for instance (i-18397eef) to become ready...
==> amazon-ebs: Waiting for SSH to become available...      #<– disconnected WiFi here
==> amazon-ebs: Timeout waiting for SSH.                    #<– made coffee here
==> amazon-ebs: Terminating the source AWS instance...      #<– turned WiFi back on once this appeared here
==> amazon-ebs: Deleting temporary security group...
==> amazon-ebs: Deleting temporary keypair...
Build 'amazon-ebs' errored: Timeout waiting for SSH.

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: Timeout waiting for SSH.

==> Builds finished but no artifacts were created.

$ echo $?
1
```

note the non-zero exit status. What am I missing here, that can help reproduce this? 

Thanks again!

**Comment by dluchinienoc** [comment]
In my particular case, the issue wasn't that Packer couldn't connect to SSH (Packer fails as expected if this happens), but rather that it's able to connect via SSH, but isn't able to authenticate via SSH.

I'm currently working on a test case with the amazon-ebs builder to make this happen.

**Comment by catsby** [comment]
>  but rather that it's able to connect via SSH, but isn't able to authenticate via SSH.

Yeah I see that... working on testing that as well, but am I picking up that the key issue was that packer itself was returning a zero exit code, and your CI interpreted that as `SUCCESS`?

**Comment by dluchinienoc** [comment]
Here's a reproducible test case using the public CentOS 6.5 AMI.

packer_1.json does nothing but take the public AMI and disable public key authentication:
```
{
  "builders": [
    {
      "access_key": "<redacted>",
      "ami_name": "Packer_2004_dluchini",
      "instance_type": "t1.micro",
      "region": "us-east-1",
      "run_tags": {
        "Name": "Packer_2004_dluchini"
      },
      "secret_key": "<redacted>",
      "source_ami": "ami-8997afe0",
      "ssh_username": "root",
      "tags": {
        "Name": "Packer_2004_dluchini"
      },
      "type": "amazon-ebs",
      "name": "amazon-ebs"
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "inline": ["sed -i 's/PubkeyAuthentication yes/PubkeyAuthentication no/' /etc/ssh/sshd_config"]
    }
  ]
}
```

packer_2.json does no provisioning at all, on top of the AMI output by packer_1.json:
```
{
  "builders": [
    {
      "access_key": "<redacted>",
      "ami_name": "Packer_2004_2_dluchini",
      "instance_type": "t1.micro",
      "region": "us-east-1",
      "run_tags": {
        "Name": "Packer_2004_dluchini"
      },
      "secret_key": "<redacted>",
      "source_ami": "<output of packer_1.json>",
      "ssh_username": "root",
      "tags": {
        "Name": "Packer_2004_2_dluchini"
      },
      "type": "amazon-ebs",
      "name": "amazon-ebs"
    }
  ]
}
```

The output of `packer build packer_2.json` is:
```
[dluchini@localhost ~/Downloads/packer]$ packer build packer_2.json
amazon-ebs output will be in this color.

==> amazon-ebs: Inspecting the source AMI...
==> amazon-ebs: Creating temporary keypair: packer 5579ea6e-bab4-46ae-2937-9ee03574fc3a
==> amazon-ebs: Creating temporary security group for this instance...
==> amazon-ebs: Authorizing SSH access on the temporary security group...
==> amazon-ebs: Launching a source AWS instance...
    amazon-ebs: Instance ID: i-9030a139
==> amazon-ebs: Waiting for instance (i-9030a139) to become ready...
==> amazon-ebs: Waiting for SSH to become available...
==> amazon-ebs: Error waiting for SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: Deleting temporary security group...
==> amazon-ebs: Deleting temporary keypair...
Build 'amazon-ebs' finished.

==> Builds finished. The artifacts of successful builds are:
[dluchini@localhost ~/Downloads/packer]$ echo $?
0
```

**Comment by catsby** [comment]
Excellent! 
Thank for the configs+steps, not to mention the time and energy :smile: 

I'm looking into this now

**Comment by m1keil** [comment]
I've reopened my issue as #2216 since it looks like two separate problems.

**Comment by catsby** [comment]
This should be fixed in #2217 

```console
$ packer build p2.json
amazon-ebs output will be in this color.

==> amazon-ebs: Prevalidating AMI Name...
==> amazon-ebs: Inspecting the source AMI...
==> amazon-ebs: Creating temporary keypair: packer 5579faff-d5b3-2505-4307-c43f83bd6832
==> amazon-ebs: Creating temporary security group for this instance...
==> amazon-ebs: Authorizing SSH access on the temporary security group...
==> amazon-ebs: Launching a source AWS instance...
    amazon-ebs: Instance ID: i-8bed7d22
==> amazon-ebs: Waiting for instance (i-8bed7d22) to become ready...
==> amazon-ebs: Waiting for SSH to become available...
==> amazon-ebs: Error waiting for SSH: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: Deleting temporary security group...
==> amazon-ebs: Deleting temporary keypair...
Build 'amazon-ebs' errored: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain

==> Builds finished but no artifacts were created.
$ echo $?
1
```

**Comment by catsby** [comment]
Merged #2217, thanks again for all the help!

**Comment by spyderdyne** [comment]
what can I do to see why my vm cant be connected to via ssh if there is no way to disable headless mode?

http://pastebin.com/B6zLxeUN

http://pastebin.com/0azhShPE

I have tried "headless": "false", and "headless": false, but it always tells me headless mode is enabled.

Thanks.
